### PR TITLE
Refactor collections

### LINKED ISSUE #46: Convert simulation box store to List
Remove getBoxCount, getBox(i) and just return the list of boxes.

### LINKED ISSUE #49: Remove util.Arrays
Convert usages to java library functions when possible, or look into primitive ArrayList libraries.

### LINKED ISSUE #50: Convert util.EnumeratedType uses to java Enums
not 2005 anymore

### LINKED ISSUE #51: Remove IEvent
Empty interface. Convert fields to concrete types.

### LINKED ISSUE #54: Look at synchronized methods
Maybe they don't have to be

### PR DESCRIPTION
**Pull request by akofke**
Another round of refactoring and cleanup, focusing on using Java Collections classes where appropriate and reworking the internals of AgentManager classes.

Closes #46, closes #49, closes #50, closes  #51, closes #54.

## AgentManagers

* Atom and Molecule agent managers have their array logic encapsulated in a custom IndexMap, which uses an array to map ints (indexes) to objects efficiently. 

* All other agent managers (Species, Box, AtomType, etc) now use HashMaps internally. These client types aren't high-churn enough for the performance of using an array to matter. In fact, all uses of these managers where there is no source and agents are just put manually to map a Species/Box/AtomType to an object are replaced with Maps.

* All agent managers now require non-null sources and relevant EventListener targets (Box, Simulation). If you're not providing a source you might as well just use a Map. 

## Collections

* Remove most methods in etomica.util.Arrays since they have equivalents in java.util.Arrays and replace their uses.

* Replace many uses of manually resized arrays with List. Things such as Boxes and Species in a Simulation and AtomTypes in a Species are now stored in an ArrayList and have methods to return the list. Iterating over this list should be preferred to using a for-i loop with getXCount and getX(i) when doing bulk operations.

* In all event managers, remove synchronization from methods and instead use a CopyOnWriteArrayList to store listeners, which is a thead-safe list meant for infrequent updating and frequent access (iteration).

## Misc

* new convenience varargs static factory method for manually creating Vectors. Instead of doing
```java
Vector v = space.makeVector(new double[]{1.0, 2.0, 3.0});
```
just use
```java
Vector v = Vector.of(1.0, 2.0, 3.0);
```
The old makeVector method is now deprecated to allow intellij to auto-change it to the new method and remove the array creation.

* Convenience static factory method for creating an AtomType with a new ElementSimple: `AtomType.simple(symbol, mass)`

* Replace most uses of EnumeratedType with Enums.

* Remove some unnecessary parameters (esp. from SimulationGraphic).

* AtomFilter no longer has an accept(molecule) method.

### CODE REVIEW

File: etomica-apps/src/main/java/etomica/AlkaneEH/SpeciesAlkaneEH.java [hunk 1]
```diff
@@ -9,83 +9,86 @@
...
-	    protected final Space space;
-	    protected final int numCarbons;
-    protected final int totalAtoms;
+    public final int numCarbons;
+    public final int numH;
+    public final int numCH2;
+    protected final Space space;
+    private final int totalAtoms;
+    private final AtomType c_3Type, c_2Type, hType;
     public int indexC_3_1;//CH3 on the left
```
Comment (ajschult) [inline etomica-apps/src/main/java/etomica/AlkaneEH/SpeciesAlkaneEH.java:30]: it looks like indexC_3_1 and _2 do not need to be fields at all

Comment (akofke) [reply etomica-apps/src/main/java/etomica/AlkaneEH/SpeciesAlkaneEH.java:30]: Do you mean they can be methods or that they don't have to exist at all? [ConformationAlkaneEH](https://github.com/etomica/etomica/blob/a4ec260974cf842bf3d1198377f1c0d4d573fda6/etomica-apps/src/main/java/etomica/virial/ConformationAlkaneEH.java#L43) uses them, although that class isn't used anywhere.

Comment (ajschult) [reply etomica-apps/src/main/java/etomica/AlkaneEH/SpeciesAlkaneEH.java:30]: Ah, indeed.  I'm pretty sure Shu used ConformationAlkaneEH to collect data; she probably just didn't commit the main class.  And I don't know why it wants the species to tell it where 2 of the carbons are but has everything else hardcoded.... oh well.

Comment (akofke) [reply etomica-apps/src/main/java/etomica/AlkaneEH/SpeciesAlkaneEH.java:30]: Ah right it already gets numCarbons from the species in the constructor. Do you think we can get the main class? If nothing else just so we don't delete this stuff on accident.

Comment (ajschult) [reply etomica-apps/src/main/java/etomica/AlkaneEH/SpeciesAlkaneEH.java:30]: er, ConformationAlkaneEH is used on 52 of this class...

Comment (akofke) [reply etomica-apps/src/main/java/etomica/AlkaneEH/SpeciesAlkaneEH.java:30]: Huh there's actually a ConformationAlkaneEH in both etomica.AlkaneEH and etomica.virial. I don't know if they're functionally the same or not.

Comment (ajschult) [reply etomica-apps/src/main/java/etomica/AlkaneEH/SpeciesAlkaneEH.java:30]: etomica.virial.ConformationAlkaneEH is very different.  In Shu's files, there is only etomica.virial.ConformationAlkaneEH_backup and etomica.virial.ConformationAlkaneEH_backkkup.  She probably committed the stuff in virial, then copied it to etomica.AlkaneEH (and committed those), but didn't cvs delete the stuff from etomica.virial.  So, let's delete the unused EH stuff etomica.virial.

**Comment by ajschult** [comment]
FYI, this appears to have a ~5% performance hit for larger systems, based on tinderbox results.
The hit is noticeable for the largest system size for both HSMD3D and SWChain and is consistent for JDK 1.8 and 1.9
http://rheneas.eng.buffalo.edu/tinderbox/showbuilds.cgi?tree=master

**Comment by akofke** [comment]
I changed the maps for AtomType and Species AgentManager to IndexMap and got some speedup with JMH, I'll commit those and see if it makes any difference on tinderbox.

**Comment by ajschult** [comment]
The IndexMap change seems to have fixed most of the hit, but the HSMD3D test (with 32000 atoms) still has a 2-3% hit that is now noticeable with more data.

http://rheneas.eng.buffalo.edu/graph/query.cgi?tbox=douglas_linux_jdk_1.9&testname=HSMD3D_32000&autoscale=1&size=&units=&ltype=&points=&showpoint=&avg=1&days=60

we went from ~89 to ~92 (~94 without the additional IndexMap change).  SWChain seems unaffected for JDK9, but other JDKs have a noticeable change there (and perhaps also for HSMD3D)

http://rheneas.eng.buffalo.edu/graph/query.cgi?tbox=douglas_linux_jdk_1.8&testname=SWChain_4000&autoscale=1&size=&units=&ltype=&points=&showpoint=&avg=1&days=60

**Pull request merged**